File: src/main/python/AoC2020_10.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 10
#

import sys
from collections import Counter
from collections import defaultdict

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST1 = """\
16
10
15
5
1
11
7
19
6
12
4
"""
TEST2 = """\
28
33
18
42
31
14
46
20
48
47
24
23
49
45
19
38
39
11
1
32
25
35
8
17
7
9
4
2
34
10
3
"""

Input = list[int]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        numbers = [0] + sorted(int(_) for _ in input_data)
        numbers.append(numbers[-1] + 3)
        return numbers

    def part_1(self, numbers: Input) -> int:
        cnt = Counter(
            numbers[i] - numbers[i - 1] for i in range(1, len(numbers))
        )
        return cnt[1] * cnt[3]

    def part_2(self, numbers: Input) -> int:
        seen = defaultdict(lambda: 0)
        seen[0] = 1
        for i, n in enumerate(numbers[1:]):
            for j in (
                numbers[k] for k in range(i, -1, -1) if n - numbers[k] <= 3
            ):
                seen[n] += seen[j]
        return seen[numbers[-1]]

    @aoc_samples(
        (
            ("part_1", TEST1, 35),
            ("part_1", TEST2, 220),
            ("part_2", TEST1, 8),
            ("part_2", TEST2, 19208),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2020, 10)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2020_10.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;
import static com.github.pareronia.aoc.IterTools.enumerate;
import static com.github.pareronia.aoc.Utils.last;
import static java.util.stream.Collectors.toCollection;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.github.pareronia.aoc.Counter;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2020_10 extends SolutionBase<List<Integer>, Long, Long> {
	
	private AoC2020_10(final boolean debug) {
		super(debug);
	}
	
	public static AoC2020_10 create() {
		return new AoC2020_10(false);
	}

	public static AoC2020_10 createDebug() {
		return new AoC2020_10(true);
	}
	
	@Override
    protected List<Integer> parseInput(final List<String> inputs) {
		final List<Integer> numbers = new ArrayList<>(List.of(0));
		inputs.stream()
		    .map(Integer::valueOf)
		    .sorted()
		    .collect(toCollection(() -> numbers));
	    numbers.add(last(numbers) + 3);
        return numbers;
    }

    @Override
	public Long solvePart1(final List<Integer> numbers) {
	    final Counter<Integer> jumps = new Counter<>(
	        range(numbers.size()).intStream().skip(1)
	            .mapToObj(i -> numbers.get(i) - numbers.get(i - 1)));
	    return jumps.get(1) * jumps.get(3);
	}
	
    // 0: 1, 1: 1, 4: 1, 5: 1, 6: 2, 7: 4, 10: 4, 11: 4, 12: 8, 15: 8, 16: 8, 19: 8
	@Override
	public Long solvePart2(final List<Integer> numbers) {
	    final Map<Integer, Long> map = new HashMap<>();
	    map.put(0, 1L);
	    enumerate(numbers.stream().skip(1)).forEach(e -> {
	        range(e.index(), -1, -1).intStream()
	                .map(numbers::get)
	                .filter(j -> e.value() - j <= 3)
                    .forEach(j -> map.merge(e.value(), map.get(j), Long::sum));
	        log(map);
	    });
		return map.get(last(numbers));
	}

	@Samples({
	    @Sample(method = "part1", input = TEST1, expected = "35"),
	    @Sample(method = "part1", input = TEST2, expected = "220"),
	    @Sample(method = "part2", input = TEST1, expected = "8"),
	    @Sample(method = "part2", input = TEST2, expected = "19208"),
	})
	public static void main(final String[] args) throws Exception {
	    AoC2020_10.create().run();
	}
	
	private static final String TEST1 = """
	        16
	        10
	        15
	        5
	        1
	        11
	        7
	        19
	        6
	        12
	        4
	        """;
	private static final String TEST2 = """
	        28
	        33
	        18
	        42
	        31
	        14
	        46
	        20
	        48
	        47
	        24
	        23
	        49
	        45
	        19
	        38
	        39
	        11
	        1
	        32
	        25
	        35
	        8
	        17
	        7
	        9
	        4
	        2
	        34
	        10
	        3
	        """;
}